### DISCUSSION TITLE
Proto files not needed for components in labs?

### DISCUSSION THREAD

**Discussion by simkimsia** [google/mesop#503] [Q&A] [answered]
I'm sorry I'm not familiar with protobuf so at the risk of asking an incredibly dumb question

I noticed the labs don't contain any proto files for content components like text to text

Does it mean that proto files are only meant for low level native components?

And not for content components?

**Comment by wwwillchen** [comment, marked as answer]
No worries. So the two main types of components (explain here: https://google.github.io/mesop/components/#types-of-components) are 1) native components (which are low-level) and use protos and 2) user-defined components, which lab components are an example of, and they don't use protos because these components are entirely written in Python.

Content components (see linked page) is an orthogonal concern and is whether the component accepts a child component. Content components can be native or user-defined.

**Reply by simkimsia** [reply]
Sorry i needed to ask a follow up question. How do we know when to use a proto file for a component?

Or is that never for user-defined components? User can also define very low level (like native components), yes?

**Reply by wwwillchen** [reply]
You don't need to use protos for user-definded components, only native components. Users can't define native components, but users can define [web components](https://google.github.io/mesop/components/#web-components) which are almost as powerful as native components; web components don't require protos.
